### PR TITLE
Ability to add multiple metrics reporters to scan

### PR DESCRIPTION
**Pull request by karuppayya**
Adds ability to add multiple reporter to the Scan

### CODE REVIEW

**Comment by karuppayya** [comment]
@aokolnychyi @RussellSpitzer @flyrain @anuragmantri @szehon-ho

File: api/src/main/java/org/apache/iceberg/BatchScanAdapter.java [hunk 1]
```diff
@@ -140,4 +140,8 @@ public int splitLookback() {
   public long splitOpenFileCost() {
     return scan.splitOpenFileCost();
   }
+
+  public Scan getScanDelegate() {
```
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/BatchScanAdapter.java:144]: it's not clear why is this change is needed. Could you elaborate please?

File: core/src/main/java/org/apache/iceberg/SnapshotScan.java [hunk 2]
```diff
@@ -66,7 +66,8 @@ protected Long snapshotId() {
 
   protected abstract CloseableIterable<T> doPlanFiles();
 
-  protected ScanMetrics scanMetrics() {
+  @Override
+  public ScanMetrics scanMetrics() {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/SnapshotScan.java:70]: `ScanMetrics` shouldn't be exposed and should be treated as internal information as they can change while the scan is going on. What you'd rather want to do is to look at the `ScanReport`, which will be created once a scan is complete.

So in your case what would make sense is to have a custom `MetricsReporter` that receives the `ScanReport` when a scan is done, which you can then use to hook it up to the Spark UI.
In order to

**Review by nastra — CHANGES_REQUESTED** [review]
We shouldn't be exposing `ScanMetrics` but rather introduce a custom `MetricsReporter` that receives a `ScanReport` once a scan is complete. This custom reporter can then add the results from the scan to the Spark UI.

File: api/src/main/java/org/apache/iceberg/BatchScanAdapter.java (hunk 1 above)
Comment (karuppayya) [reply api/src/main/java/org/apache/iceberg/BatchScanAdapter.java:144]: SnapShotScan has the relevant read metrics. Exposing the backing scan object from the adapter to extract the metrics. (which will be used [here](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/Scan.java#L126))

File: api/src/main/java/org/apache/iceberg/Table.java [hunk 3]
```diff
@@ -48,6 +49,18 @@ default String name() {
...
 
+  /**
+   * Create a new {@link TableScan scan} for this table with a {@link MetricsReporter metrics
+   * reporter}.
+   *
+   * <p>Once a table scan is created, it can be refined to project columns and filter data.
+   *
+   * @return a table scan for this table
+   */
+  default TableScan newScan(MetricsReporter metricsReporter) {
```
Comment (aokolnychyi) [inline api/src/main/java/org/apache/iceberg/Table.java:60]: Instead of modifying `Table`, we should probably extend `Scan` to accept any number of custom metrics report in addition to the default one. It should behave like a listener pattern.

File: api/src/main/java/org/apache/iceberg/Scan.java [hunk 4]
```diff
@@ -171,4 +172,7 @@ default ThisT select(String... columns) {
 
   /** Returns the split open file cost for this scan. */
   long splitOpenFileCost();
+
+  /** Create a new scan that will report the scan metrics to the {@code reporter} */
+  ThisT withMetricsReporter(Collection<MetricsReporter> reporter);
```
Comment (aokolnychyi) [inline api/src/main/java/org/apache/iceberg/Scan.java:177]: I don't think we use `with` prefixes throughout the `Scan` API. Also, we should accept a single reporter rather than a collection given the method name. Under the hood, it would be a list and we should be able to call this method more than once to add different reporters.

```
scan
    .filter(cold)
    .metricsReporter(reporter1)
    .metricsReporter(reporter2)
    .planFiles()
```

We may add a new method that accepts an iterable in the future if we have a use case for it.

File: core/src/main/java/org/apache/iceberg/TableScanContext.java [hunk 5]
```diff
@@ -351,14 +352,15 @@ TableScanContext planWith(ExecutorService executor) {
         toSnapshotId,
         executor,
         fromSnapshotInclusive,
-        metricsReporter);
+        metricsReporters);
   }
 
-  MetricsReporter metricsReporter() {
-    return metricsReporter;
+  Collection<MetricsReporter> metricsReporter() {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/TableScanContext.java:358]: Should the method name change too?

File: core/src/main/java/org/apache/iceberg/TableScanContext.java [hunk 6]
```diff
@@ -351,14 +352,15 @@ TableScanContext planWith(ExecutorService executor) {
...
+    return metricsReporters;
   }
 
   TableScanContext reportWith(MetricsReporter reporter) {
+    metricsReporter().add(reporter);
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/TableScanContext.java:363]: If I remember correctly, this class is immutable and we create a new instance on every call. We should probably follow what is done for updating properties in `withOption`, where we copy existing options into a new list before adding.

```
TableScanContext reportWith(MetricsReporter reporter) {
  ImmutableList.Builder<MetricsReporter> builder = ImmutableList.builder();
  builder.addAll(metricsReporters);
  builder.add(reporter);
  List<MetricsReporter> newMetricsReporters = builder.build();
  return new TableScanContext(
      snapshotId,
      rowFilter,
      ignoreResiduals,
      caseSensitive,
      colStats,
      projectedSchema,
      selectedColumns,
      options,
      fromSnapshotId,
      toSnapshotId,
      planExecutor,
      fromSnapshotInclusive,
      newMetricsReporters);
}
```

File: core/src/main/java/org/apache/iceberg/TableScanContext.java [hunk 7]
```diff
@@ -372,6 +374,24 @@ TableScanContext reportWith(MetricsReporter reporter) {
         toSnapshotId,
         planExecutor,
         fromSnapshotInclusive,
-        reporter);
+        metricsReporter());
+  }
+
+  TableScanContext reportWith(Collection<MetricsReporter> reporters) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/TableScanContext.java:380]: I think it is OK to just have a method that accepts one reporter for now.

File: .palantir/revapi.yml [hunk 8]
```diff
@@ -385,10 +388,6 @@ acceptedBreaks:
       old: "method void org.apache.iceberg.SnapshotProducer<ThisT>::validate(org.apache.iceberg.TableMetadata)\
         \ @ org.apache.iceberg.StreamingDelete"
       justification: "Removing deprecations for 1.2.0"
-    - code: "java.method.returnTypeChangedCovariantly"
```
Comment (aokolnychyi) [inline .palantir/revapi.yml:388]: Why is this changed?

File: .palantir/revapi.yml [hunk 9]
```diff
@@ -66,6 +66,9 @@ acceptedBreaks:
       old: "method void org.apache.iceberg.io.DataWriter<T>::add(T)"
       justification: "Removing deprecated method"
   "1.1.0":
+    org.apache.iceberg:iceberg-api:
+    - code: "java.method.addedToInterface"
+      justification: "Add metricsreporter to Scan"
```
Comment (aokolnychyi) [inline .palantir/revapi.yml:71]: Instead of modifying checks, we should throw `UnsupportedOperationException` in the default implementation. Just like we do in some methods in `TableScan` today.

File: core/src/main/java/org/apache/iceberg/SnapshotScan.java [hunk 10]
```diff
@@ -144,7 +144,9 @@ public CloseableIterable<T> planFiles() {
                   .scanMetrics(ScanMetricsResult.fromScanMetrics(scanMetrics()))
                   .metadata(metadata)
                   .build();
-          context().metricsReporter().report(scanReport);
+          context()
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/SnapshotScan.java:147]: nit: What about an explicit for each in this case? I am not sure such statements are easy to read since they are split on multiple lines.

```
for (MetricsReporter reporter : context().metricsReporters()) {
  reporter.report(scanReport);
}
```

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkMetricsReporter.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+public class SparkMetricsReporter implements MetricsReporter {
+
+  private MetricsReport metricsReport;
+
+  @Override
+  public void report(MetricsReport report) {
+    this.metricsReport = report;
+  }
+
+  MetricsReport getMetricsReport() {
```
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkMetricsReporter.java:34]: nit: We don't use `getXXX` prefixes for getters, just `metricsReport`.

I also don't think this reporter is specific to Spark. We may think of a good name and put it in `core`. I believe it would be a common use case to intercept scan reports.

File: api/src/main/java/org/apache/iceberg/BatchScanAdapter.java (hunk 1 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/BatchScanAdapter.java]: I believe this no longer applies, let me resolve this thread.

File: core/src/main/java/org/apache/iceberg/SnapshotScan.java (hunk 2 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/SnapshotScan.java]: I believe this no longer applies too and the PR has been updated to support custom reporters.

File: .palantir/revapi.yml (hunk 9 above)
Comment (nastra) [reply .palantir/revapi.yml]: +1 to that

File: api/src/main/java/org/apache/iceberg/Scan.java (hunk 4 above)
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/Scan.java]: +1 to that suggestion

**Review by nastra — COMMENTED** [review]
@karuppayya it's a bit unclear to me whether you're planning to do the metrics reporting  to Spark as part of this PR or not. Could you clarify please?

**Comment by karuppayya** [comment]
> I believe @karuppayya meant to report from here directly to Spark metrics. @karuppayya were you planning to add that functionality as part of this PR?

Yes, that was the idea.To access the metrics in SparkScanBuilder to use [here](https://github.com/apache/spark/commit/d4c58159925133771d305cc7ac4f1248f215812c)

**Comment by karuppayya** [comment]
I am not sure of the reason for workflow needing approval to start. @RussellSpitzer @aokolnychyi @nastra @rdblue any idea why this would happen, any recent change?

**Comment by aokolnychyi** [comment]
Since we don't know how our metrics reporting will look like until we support Spark 3.4, what about focusing only on the core logic for adding custom reporters?

File: api/src/main/java/org/apache/iceberg/Scan.java [hunk 12]
```diff
@@ -171,4 +172,10 @@ default ThisT select(String... columns) {
 
   /** Returns the split open file cost for this scan. */
   long splitOpenFileCost();
+
+  /** Create a new scan that will report the scan metrics to the {@code reporter} */
+  default ThisT metricsReporter(MetricsReporter reporter) {
+    throw new UnsupportedOperationException(
+        this.getClass().getName() + " doesn't implement metricReporter");
```
Comment (aokolnychyi) [inline api/src/main/java/org/apache/iceberg/Scan.java:179]: Typo in the method name inside the comment (should be `metricsReporter`).

File: api/src/main/java/org/apache/iceberg/Scan.java [hunk 13]
```diff
@@ -171,4 +172,10 @@ default ThisT select(String... columns) {
 
   /** Returns the split open file cost for this scan. */
   long splitOpenFileCost();
+
+  /** Create a new scan that will report the scan metrics to the {@code reporter} */
```
Comment (aokolnychyi) [inline api/src/main/java/org/apache/iceberg/Scan.java:176]: The doc should indicate that this adds a reporter to the list of existing reporters, not overrides it.

File: .palantir/revapi.yml [hunk 14]
```diff
@@ -67,45 +67,9 @@ acceptedBreaks:
       justification: "Removing deprecated method"
   "1.1.0":
     org.apache.iceberg:iceberg-core:
-    - code: "java.class.noLongerImplementsInterface"
```
Comment (aokolnychyi) [inline .palantir/revapi.yml:70]: Why are these changes needed?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 15]
```diff
@@ -420,12 +421,14 @@ private Scan buildBatchScan() {
   private Scan buildBatchScan(Long snapshotId, Long asOfTimestamp, String branch, String tag) {
     Schema expectedSchema = schemaWithMetadataColumns();
 
+    SparkMetricsReporter reporter = new SparkMetricsReporter();
```
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:424]: I don't think there is any value in this class at this point. Let's remove Spark classes from this PR and add them when we know how the reporting logic will look like.

File: core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java [hunk 16]
```diff
@@ -42,6 +43,33 @@ public TestScanPlanningAndReporting() {
     super(2);
   }
 
+  @Test
+  public void scanningWithMutipleReporters() throws IOException {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java:47]:
```suggestion
  public void scanningWithMultipleReporters() throws IOException {
```

File: core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java [hunk 17]
```diff
@@ -42,6 +43,33 @@ public TestScanPlanningAndReporting() {
...
+                  reportedCount.getAndIncrement();
+                })
+            .metricsReporter(
+                (MetricsReporter) -> {
+                  reportedCount.getAndIncrement();
+                });
+    try (CloseableIterable<FileScanTask> fileScanTasks = tableScan.planFiles()) {
+      fileScanTasks.forEach(task -> {});
+    }
+    assertThat(reportedCount.get()).isEqualTo(2);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java:70]:
```suggestion
    try (CloseableIterable<FileScanTask> fileScanTasks = tableScan.planFiles()) {
      fileScanTasks.forEach(task -> {});
    }

    assertThat(reportedCount.get()).isEqualTo(2);

    // make sure default metrics reporter is still reporting
    ScanReport scanReport = reporter.lastReport();
    assertThat(scanReport).isNotNull();
    assertThat(scanReport.tableName()).isEqualTo(tableName);
    assertThat(scanReport.snapshotId()).isEqualTo(1L);
    ScanMetricsResult result = scanReport.scanMetrics();
    assertThat(result.totalPlanningDuration().totalDuration()).isGreaterThan(Duration.ZERO);
    assertThat(result.resultDataFiles().value()).isEqualTo(1);
```

Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java:70]: this is also to make sure the default metrics reporter still reports

**Review by nastra — COMMENTED** [review]
almost there, just a few small things to address and then I think we can get this merged

File: api/src/main/java/org/apache/iceberg/Scan.java (hunk 12 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/Scan.java:182]: Looks like this comment was missed. The exception message should include the correct method name.

File: api/src/main/java/org/apache/iceberg/Scan.java [hunk 18]
```diff
@@ -171,4 +172,13 @@ default ThisT select(String... columns) {
 
   /** Returns the split open file cost for this scan. */
   long splitOpenFileCost();
+
+  /**
+   * Create a new scan that will report the scan metrics to the {@code reporter} {@code reporter} is
```
Comment (aokolnychyi) [inline api/src/main/java/org/apache/iceberg/Scan.java:177]: nit: Duplicate `{@code reporter}`?

Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/Scan.java:177]: I'd actually consider adapting the message as below.

```
Create a new scan that will report scan metrics to the provided reporter in addition to reporters maintained by the scan.
```

**Review by aokolnychyi — APPROVED** [review]
Two minor nits and should be good to go.

**Review by nastra — APPROVED** [review]
LGTM once the two nits are fixed. Also we should probably adjust the PR title / commit msg when merging to better reflect the scope of this work

**Comment by aokolnychyi** [comment]
Thank you, @karuppayya! Thanks for reviewing, @nastra!
